src/coins.h

```diff
@@ -127,9 +151,49 @@ struct CCoinsCacheEntry
         FRESH = (1 << 1),
     };
 
-    CCoinsCacheEntry() : flags(0) {}
-    explicit CCoinsCacheEntry(Coin&& coin_) : coin(std::move(coin_)), flags(0) {}
-    CCoinsCacheEntry(Coin&& coin_, unsigned char flag) : coin(std::move(coin_)), flags(flag) {}
+    //! Adding a flag also requires a self reference to the pair that contains
+    //! this entry in the CCoinsCache map and a reference to the head of the
+    //! flagged pair linked list.
+    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& head)
+    {
+        if (!m_flags && flags) {
+            m_prev = &head;
+            m_next = head.second.m_next;
+            if (m_next) m_next->second.m_prev = &self;
+            head.second.m_next = &self;
```

Comment:
I'm still thinking about this, it's basically a static method that doesn't need `this`, which seems to indicate to me that it might be misplaced:
```
    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& head)
    {
        if (!m_flags && flags) {
            AddToList(self, head);
        }
        m_flags |= flags;
    }
    static inline void AddToList(CoinsCachePair& self, CoinsCachePair& head) {
        self.second.m_prev = &head;
        self.second.m_next = head.second.m_next;
        if (head.second.m_next) head.second.m_next->second.m_prev = &self;
        head.second.m_next = &self;
    }
```
No recommendation yet, just thinking out loud...